src/porepy/compositional/base.py

```diff
@@ -0,0 +1,855 @@
+
+        Returns:
+            A list of molality values per pseudo-component.
+
+        """
+        if len(fractions) != len(self._pseudo_comps):
+            raise ValueError(
+                f"Need {len(self._pseudo_comps)} values, {len(fractions)} given."
+            )
+
+        molalities = []
+
+        # solvent fraction
+        x_s = 1 - safe_sum(fractions)
+        for fraction in fractions:
+            m_i = fraction / (x_s * self.molar_mass)
+            molalities.append(m_i)
+
+        return molalities
+
+    def fractions_from_molalities(self, *molalities: FloatType) -> list[FloatType]:
+        """Reverse operation for :meth:`molalities_from_fractions`.
+
+        Parameters:
+            *molalities: Molalities per present pseudo-component.
+
+        Raises:
+            ValueError: If the number of provided values does not match the number
+                of present pseudo-components.
+
+        Returns:
+            A list of relative solute fractions calculated from molalities.
+
+        """
+        if len(molalities) != len(self._pseudo_comps):
+            raise ValueError(
+                f"Need {len(self._pseudo_comps)} values, {len(molalities)} given."
+            )
+
+        m_sum = safe_sum(molalities)
+        X = list()
+
+        for m in molalities:
+            # See https://en.wikipedia.org/wiki/Molality
+            x_i = self.molar_mass * m / (1 + self.molar_mass * m_sum)
+            X.append(x_i)
+
+        return X
+
+
+class AbstractEoS(abc.ABC):
+    """Abstract EoS class defining the interface between thermodynamic input
+    and resulting structure containing thermodynamic properties of a phase.
+
+    Component properties required for computations can be extracted in the constructor.
+
+    Note:
+        This class is called 'abstract EoS'. Users can implement any correlations but
+        are encouraged to focus on thermodynamic consistency.
+
+        Phase properties are defined as secondary expressions, and the framework is
+        able to pick up the dependencies and call :meth:`compute_phase_state` with
+        the right values.
+
+        For more information on this, see
+        :class:`~porepy.compositional.compositional_mixins.FluidMixtureMixin` and
+        :meth:`~porepy.compositional.compositional_mixins.FluidMixtureMixin.
+        dependencies_of_phase_properties`.
+
+    Parameters:
+        components: A sequence of components for which the EoS is instantiated.
+
+    Raises:
+        CompositionalModellingError: If no components passed.
+
+    """
+
+    def __init__(self, components: Sequence[Component]) -> None:
+        self._nc: int = len(components)
+        """Number of components passed at instantiation."""
+
+        if self._nc == 0:
+            raise CompositionalModellingError("Cannot create an EoS with no components")
+
+    @abc.abstractmethod
+    def compute_phase_state(
+        self, phase_type: int, *thermodynamic_input: np.ndarray
```

Comment:
I thought of including scalars, but that would be hell for the coming parallelized numba computations.
So I left only numpy arrays, also since this is intended to represent a fluid on a grid with cells.

For single-cell values, the user would have to wrap the values into arrays of shape ``(1,)``.

AdArrays are not required, since the computations of derivatives are not implemented using AD, but explicitly.

Reply:
Thanks for the clarification - I trust your judgement here.